### PR TITLE
Add support for hier_path using new AST_DOT

### LINKED ISSUE #228: Earlgrey: yosys doesn't support in array of typedef in struct.
Yosys currently doesn't support arrays of typedef in struct, we need to add support for them. Example test:
```
module dut();
typedef struct packed {
  logic        q;
} pwrmgr_reg2hw_wakeup_en_mreg_t;
typedef struct packed {
  logic        q;
} pwrmgr_reg2hw_intr_state_reg_t;
typedef struct packed {
  pwrmgr_reg2hw_intr_state_reg_t intr_state;
  
} pwrmgr_reg2hw_t;

pwrmgr_reg2hw_t t;
pwrmgr_reg2hw_wakeup_en_mreg_t [15:0] SECOND;
endmodule
```

This can be patched by changing ``typedef`` to exact type, e.g.:
``pwrmgr_reg2hw_wakeup_en_mreg_t [15:0] wakeup_en;`` to ``logic [15:0] wakeup_en;``

### PR DESCRIPTION
**Pull request by kamilrakoczy**
Fixes: https://github.com/chipsalliance/UHDM-integration-tests/issues/228

Corresponding uhdm-plugin PR: https://github.com/antmicro/yosys-symbiflow-plugins/pull/34

### CODE REVIEW

File: frontends/ast/simplify.cc [hunk 1]
```diff
@@ -268,6 +263,106 @@ static AstNode *make_range(int left, int right, bool is_signed = false)
 	return range;
 }
 
+static AstNode *expand_dot(const AstNode* current_struct, const AstNode* search_node) {
+	AstNode *current_struct_elem = nullptr;
+	for (auto c : current_struct->children) {
```
Comment (kbieganski) [inline frontends/ast/simplify.cc]: Can't this be replaced with a `vector.find`?
`search_str` should be defined before this loop.

File: frontends/ast/simplify.cc [hunk 2]
```diff
@@ -268,6 +263,106 @@ static AstNode *make_range(int left, int right, bool is_signed = false)
...
+		right = AstNode::mkconst_int(current_struct_elem->range_right, true);
+	} else if (current_struct_elem->type == AST_STRUCT) {
+		// Struct can have multiple range, so to get size of 1 struct,
+		// we get left range for first children, and right range for last children
+		left = AstNode::mkconst_int(current_struct_elem->children.front()->range_left, true);
+		right = AstNode::mkconst_int(current_struct_elem->children.back()->range_right, true);
+	} else {
+		// Structs currently can only have AST_STRUCT or AST_STRUCT_ITEM
+		// so, it should never happend
+		log_assert(1 == 0);
```
Comment (kbieganski) [inline frontends/ast/simplify.cc]: I know this shouldn't happen, but we should have a readable error message here.
Also, replace all instances of `never happend` with `never happen`.

File: frontends/ast/simplify.cc [hunk 3]
```diff
@@ -268,6 +263,106 @@ static AstNode *make_range(int left, int right, bool is_signed = false)
...
+				right = new AstNode(AST_ADD, right, mul->clone());
+			} else {
+				log_assert(1 == 0); // should never happend
+			}
+		} else {
+			log_assert(1 == 0); // should never happend
+		}
+	}
+	// Return range from the begining of *current* struct
+	// When all AST_DOT will be expanded it will return range
```
Comment (kbieganski) [inline frontends/ast/simplify.cc]: _When all AST_DOT **are** expanded_

File: frontends/ast/simplify.cc [hunk 4]
```diff
@@ -2199,100 +2318,98 @@ bool AstNode::simplify(bool const_fold, bool at_zero, bool in_lvalue, int stage,
...
+			} else if (id2ast->type == AST_CELL || id2ast->type == AST_INTERFACEPORT || id2ast->type == AST_AUTOWIRE) {
+				// for accessing elements currently unsupported with AST_DOT
+				// fallback to "." notation
+				str += "." + (*dot_ast)->str.substr(1);
+				children.clear(); // TODO: memory leak here
+				return true;
+			} else {
+				id2ast->dumpAst(NULL, "id2ast >");
+				dumpAst(NULL, "this >");
+				log_assert(1 == 0);
```
Comment (kbieganski) [inline frontends/ast/simplify.cc:2345]: What's the error here? Should be a readable message

**Review by kbieganski — CHANGES_REQUESTED** [review]

File: frontends/ast/simplify.cc (hunk 1 above)
Comment (kamilrakoczy) [reply frontends/ast/simplify.cc]: Done.

File: frontends/ast/simplify.cc (hunk 2 above)
Comment (kamilrakoczy) [reply frontends/ast/simplify.cc]: Done.

File: frontends/ast/simplify.cc (hunk 3 above)
Comment (kamilrakoczy) [reply frontends/ast/simplify.cc]: Fixed.

File: frontends/ast/simplify.cc (hunk 4 above)
Comment (kamilrakoczy) [reply frontends/ast/simplify.cc:2345]: Added.

**Comment by kamilrakoczy** [comment]
I fixed review comments and also squashed commits.

**Review by kbieganski — APPROVED** [review]

**Pull request merged**